posthog/temporal/feature_flag_sync/workflow.py

```diff
@@ -0,0 +1,219 @@
+import json
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+
+from django.utils import timezone as django_timezone
+
+import structlog
+import temporalio.common
+import temporalio.activity
+import temporalio.workflow
+
+from posthog.clickhouse.client import sync_execute
+from posthog.exceptions_capture import capture_exception
+from posthog.models.feature_flag.feature_flag import FeatureFlag
+from posthog.redis import get_client
+from posthog.temporal.common.base import PostHogWorkflow
+from posthog.temporal.feature_flag_sync.metrics import (
+    get_feature_flag_sync_duration_metric,
+    get_feature_flag_sync_events_processed_metric,
+    get_feature_flag_sync_finished_metric,
+    get_feature_flag_sync_flags_updated_metric,
+)
+
+logger = structlog.get_logger(__name__)
+
+FEATURE_FLAG_LAST_CALLED_SYNC_KEY = "posthog:feature_flag_last_called_sync:last_timestamp"
+
+
+@dataclass
+class SyncFeatureFlagLastCalledInputs:
+    """Inputs for syncing feature flag last_called_at timestamps."""
+
+    # Could add parameters like batch_size, lookback_hours etc. in the future
+    pass
+
+
+@dataclass
+class FlagSyncResult:
+    """Result of syncing feature flag timestamps."""
+
+    updated_count: int
+    processed_events: int
+    sync_duration_seconds: float
+
+
+@temporalio.activity.defn
+def sync_feature_flag_last_called_activity(inputs: SyncFeatureFlagLastCalledInputs) -> FlagSyncResult:
+    """
+    Sync last_called_at timestamps from ClickHouse $feature_flag_called events to PostgreSQL.
+
+    This activity:
+    1. Gets the last sync timestamp from Redis
+    2. Queries ClickHouse for flag usage since last sync
+    3. Bulk updates PostgreSQL with latest timestamps
+    4. Updates the sync checkpoint in Redis
+    """
+    start_time = django_timezone.now()
+
+    try:
+        redis_client = get_client()
+
+        last_sync_timestamp = get_last_sync_timestamp(redis_client, lookback_days=1)
+        current_sync_timestamp = django_timezone.now()
+
+        logger.info(
+            "Starting feature flag sync",
+            last_sync_timestamp=last_sync_timestamp.isoformat(),
+            current_sync_timestamp=current_sync_timestamp.isoformat(),
+        )
+
+        # Query ClickHouse for flag usage since last sync
+        # Limit 100000 for insurance against large datasets and memory issues during a surge
```

Comment:
do we think it's worth batching transactions here just in case? might help us even more with insurance.

Not blocking, just figured I'd bring it up.